### ISSUE TITLE
Cogwheels disappeared

### ISSUE THREAD

**Issue by romainpoirier** [mmikkel/CpFieldInspect-Craft#8]
Cogwheels disappeared from every page, so I can't use the plugin's shortcut anymore.

I tried to revert back to the 1.0.4 version of the plugin, and also from craftcms 3.0.27.1 to 3.0.27 and also to 3.0.26, but cogwheels are still disappeared.

I see nothing specific in the Craft logs, no error in the JS console, and I don't see the cogwheels when inspecting the HTML code.

Any one experimenting the same issue?

**Comment by mmikkel** [comment]
Hi @romainpoirier – you're _probably_ running into the same issue as a few other people have; you've got another plugin (or a custom module) that is somehow blocking CP Field Inspect's (and possibly other plugins', as well) JS and CSS from loading (see https://github.com/mmikkel/CpFieldInspect-Craft/issues/7).

In other words, I can't reproduce. Things to try:

- Check your browser's dev tools console for any JS errors
- Check your browser's dev tools' network tab to see if the files **cpfieldinspect.js** and **cpfieldinspect.css** are in fact requested and downloaded by the browser, or if they're missing
- Disable (not uninstall! and be sure to back up your database) all other plugins. Do it one by one, check if the cogwheels appear after deactivating each one. This way, if another plugin is the cause for the issue, you'll know which one.
- Remove any custom modules you might have bootstrapped in `/config/app.php`; also one by one
- Clear your Control Panel Resources

_If it does turn out to be another, publicly available plugin causing the issue w/ CP Field Inspect, I'd love to know which one._

**Comment by romainpoirier** [comment]
Thank you!

> Check your browser's dev tools console for any JS errors

Nothing else than `Double-instantiating a menu button on an element `.

> Check your browser's dev tools' network tab to see if the files cpfieldinspect.js and cpfieldinspect.css are in fact requested and downloaded by the browser, or if they're missing

They're both missing.

> Clear your Control Panel Resources

Done.

> Remove any custom modules you might have bootstrapped in /config/app.php; also one by one

This test solved the problem, but I need to use my custom module. Do you have any idea of what could interact with CpFieldInspect, and how to solve this issue?

As perhaps like in this issue ([#7](https://github.com/mmikkel/CpFieldInspect-Craft/issues/7)), I'm using some permissions tests in my custom module:
`!$currentUser->isInGroup(self::GROUP_ADMIN) && !$currentUser->admin && $currentUser->id != $entry->authorId`

**Comment by mmikkel** [comment]
@romainpoirier It sounds like a related issue as #7, for sure.

If you're able to share your custom module with me, I can do some debugging. I'll get in touch with you on the Craft Slack :)

**Comment by romainpoirier** [comment]
@mmikkel, I'm sorry, my module is still under development and I'm not allowed to share the code.
If you have any clues about what I could fix by myself, or if you would like that I send you code fragments, let me know. Thank you!

**Comment by mmikkel** [comment]
Anything you _could_ share (in terms of your custom module's code) would be helpful :)

I've tried adding `Craft::$app->getUser()->getIsAdmin()` and stuff like calls to `isInGroup()` to a custom module's `__construct()` and `init()` methods, and calling those methods doesn't, by itself, prevent CP Field Inspect's JS from loading on my end – so there must be something else going on in your module.

At the very least, it'd be interesting to take a look at your module's main class' `__construct()` and `init()` methods.

Feel free to DM me (`@mmikkel`) on the official Craft Slack if you don't want to share anything publicly here on GitHub.

**Comment by romainpoirier** [comment]
Thank you, I found the problem thanks to your advises.

Assigning this variable is causing this issue:
`$currentUser = Craft::$app->getUser()->getIdentity();`

So I changed all `$currentUser` to `Craft::$app->getUser()->getIdentity()` and that's it.

Any idea why it was blocking the CpFieldInspect plugin?

**Comment by mmikkel** [comment]
@romainpoirier Thanks for the follow-up!

Very odd; I tried that exact statement – `$currentUser = Craft::$app->getUser()->getIdentity();` – in a custom module, and it didn't prevent the cogwheels from displaying on my end. Also *super weird* that _the name of a variable_ inside a module's class method should trigger this behaviour.

Where did you have the `$currentUser = Craft::$app->getUser()->getIdentity();` in your module? In the constructor, the `init()` method or inside an event handler (if so, which one?).

**Comment by romainpoirier** [comment]
Yes, super weird! The `$currentUser = Craft::$app->getUser()->getIdentity();` was at the beginning of the `init()` method, just after this:
```
parent::init();
self::$instance = $this;
```

**Comment by mmikkel** [comment]
@romainpoirier Cool, thank you. I'll do some more testing, but if I'm not able to reproduce it I'll think we'll have to mark it down as "magic" :)

**Comment by romainpoirier** [comment]
Thank you for your help. If you're still not able to reproduce, let me know and I'll send you more codes to reproduce the context.

**Comment by mmikkel** [comment]
Here's the `init` method from my custom module (scaffolded via pluginfactory.io; nothing else added to it):

```php
    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();
        self::$instance = $this;

        $currentUser = Craft::$app->getUser()->getIdentity();

        Craft::info(
            Craft::t(
                'my-module',
                '{name} module loaded',
                ['name' => 'My Module']
            ),
            __METHOD__
        );
    }
```

Bootstrapping the module containing the above does **not** prevent CP Field Inspect's cogwheels from displaying (Craft 3.0.27.1), so I'm stumped as to why you're having this issue - there's got to be more to it. I'm still not convinced the _name_ of the local `$currentUser` variable is the actual cause, don't really see how that would even be possible – I could be wrong though :)

**Comment by romainpoirier** [comment]
I have made a reduced test case, deleting everything in the module except the base NameModule.php file. I finally found that the issue could be linked to the `__construct` function, which is not in your custom module.

I have cleanup the code that I can share you:

```
<?php

namespace modules\mymodule;

use Craft;
use craft\i18n\PhpMessageSource;
use yii\base\Module;

class mymodule extends Module
{
    public static $instance;

    public function __construct($id, $parent = null, array $config = [])
    {
        Craft::setAlias('@modules/mymodule', $this->getBasePath());
        $this->controllerNamespace = 'modules\mymodule\controllers';

        // Translation category
        $i18n = Craft::$app->getI18n();
        if (!isset($i18n->translations[$id]) && !isset($i18n->translations[$id.'*'])) {
            $i18n->translations[$id] = [
                'class' => PhpMessageSource::class,
                'sourceLanguage' => 'en-US',
                'basePath' => '@modules/mymodule/translations',
                'forceTranslation' => true,
                'allowOverrides' => true,
            ];
        }

        // Set this as the global instance of this module class
        static::setInstance($this);

        parent::__construct($id, $parent, $config);
    }

    public function init()
    {
        parent::init();
        self::$instance = $this;

        $currentUser = Craft::$app->getUser()->getIdentity();

        Craft::info(
            Craft::t(
                'my-module',
                '{name} module loaded',
                ['name' => 'My Module']
            ),
            __METHOD__
        );
    }
}
```

If you fit it to your module, you should now see that the cogwheels disappeared. Then you can to comment the `__construct` function, or `$currentUser = Craft::$app->getUser()->getIdentity();` to see what's happening.

**Comment by mmikkel** [comment]
@romainpoirier Awesome, thank you. I'll test it ASAP and get back to you :)

**Comment by mmikkel** [comment]
Hi @romainpoirier, sorry for the delay.

I'm able to reproduce this now, and I believe the issue is that we're actually running into this bug: https://github.com/craftcms/cms/issues/2473

...in that the reason for the cogwheels not appearing, is that `Craft::$app->getUser()->getIsAdmin()` returns false in CP Field Inspect's `init` method, if `Craft::$app->getUser()->getIdentity();` was previously called inside a custom module (phew).

Will keep you posted!

**Comment by mmikkel** [comment]
@romainpoirier OK, so I don't have a real fix – modules calling `getIdentity()` on load and causing the issue mentioned above isn't really something I _can_ account for in CP Field Inspect – but I've got a workaround that might do the trick for you:

```php
public function init()
{
    parent::init();
    self::$instance = $this;

    // Defer further boot-up until after plugins have loaded
    Event::on(
        Plugins::class,
        Plugins::EVENT_AFTER_LOAD_PLUGINS,
        function () {
            $this->doIt();
        }
    );

    Craft::info(
        Craft::t(
            'my-module',
            '{name} module loaded',
            ['name' => 'My Module']
        ),
        __METHOD__
    );
}

protected function doIt()
{
    $currentUser = Craft::$app->getUser()->getIdentity();
    // ... all other logic dependant on `$currentUser`
}
```

To explain; the above simply takes the `Craft::$app->getUser()->getIdentity()` call and rather than call it directly from the module's `init()` method, defers it to an event handler called on `Plugins::EVENT_AFTER_LOAD_PLUGINS`.

Modules load before plugins in Craft; this fix allows plugins such as CP Field Inspect to load before you call the `getIdentity()` method; circumventing the issue mentioned in [craftcms/cms#2473](https://github.com/craftcms/cms/issues/2473). As far as I've been able to test it, this fixes the issue.

Since this is a semi-common issue people are running into, I'll add the above example to the README.md.

Would be fun to hear if this works for you; if not I think you'll have to lean on P&T to take a second look at that bug - modules calling `getIdentity()` on load isn't really something that I can account for in CP Field Inspect, unfortunately. Closing this issue :)
